sql-plugin/src/main/scala/com/nvidia/spark/rapids/NvcompLZ4CompressionCodec.scala

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import ai.rapids.cudf.{BaseDeviceMemoryBuffer, ContiguousTable, Cuda, DeviceMemoryBuffer}
+import ai.rapids.cudf.nvcomp.{BatchedLZ4Compressor, BatchedLZ4Decompressor, CompressionType, Decompressor, LZ4Compressor}
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.format.{BufferMeta, CodecType}
+
+/** A table compression codec that uses nvcomp's LZ4-GPU codec */
+class NvcompLZ4CompressionCodec extends TableCompressionCodec with Arm {
+  override val name: String = "nvcomp-LZ4"
+  override val codecId: Byte = CodecType.NVCOMP_LZ4
+
+  override def compress(
+      tableId: Int,
+      contigTable: ContiguousTable): CompressedTable = {
+    val tableBuffer = contigTable.getBuffer
+    val (compressedSize, oversizedBuffer) = NvcompLZ4CompressionCodec.compress(tableBuffer)
+    closeOnExcept(oversizedBuffer) { oversizedBuffer =>
+      val tableMeta = MetaUtils.buildTableMeta(
+        tableId,
+        contigTable.getTable,
+        tableBuffer,
+        CodecType.NVCOMP_LZ4,
+        compressedSize)
+      CompressedTable(compressedSize, tableMeta, oversizedBuffer)
+    }
+  }
+
+  override def decompressBuffer(
+      outputBuffer: DeviceMemoryBuffer,
+      outputOffset: Long,
+      outputLength: Long,
+      inputBuffer: DeviceMemoryBuffer,
+      inputOffset: Long,
+      inputLength: Long): Unit = {
+    withResource(outputBuffer.slice(outputOffset, outputLength)) { outSlice =>
+      withResource(inputBuffer.slice(inputOffset, inputLength)) { inSlice =>
+        NvcompLZ4CompressionCodec.decompress(outSlice, inSlice)
+      }
+    }
+  }
+
+  override def createBatchCompressor(maxBatchMemoryBytes: Long): BatchedTableCompressor = {
+    new BatchedNvcompLZ4Compressor(maxBatchMemoryBytes)
+  }
+
+  override def createBatchDecompressor(maxBatchMemoryBytes: Long): BatchedBufferDecompressor = {
+    new BatchedNvcompLZ4Decompressor(maxBatchMemoryBytes)
+  }
+}
+
+object NvcompLZ4CompressionCodec extends Arm {
+  // TODO: Make this a config?
```

Comment:
do we want to do something with this TODO?  does the chunk size make much difference (if you had time to test)?

Reply:
I have not had time to test different chunk sizes.  Changing the chunk size will affect compression ratios and compression performance, but I don't know to what degree.